### DISCUSSION TITLE
File name and line number when the application terminates

### DISCUSSION THREAD

**Discussion by lauler1** [hsutter/cppfront#798] [Suggestions]
For debugging, it could be highly beneficial to display the file name and line number when the application terminates due to an assertion failure, contract violation, or out-of-bounds access. This would provide more informative feedback than a generic message. For example, instead of a message like:

```bash
Bounds safety violation: out-of-bounds access attempt detected
terminate called without an active exception
Aborted
```

It would be more helpful to have a message such as:

```bash
Bounds safety violation in file.cpp:123: out-of-bounds access attempt detected
terminate called without an active exception
Aborted
```

Is there an existing mechanism to achieve this?

**Comment by JohelEGP** [comment]
From `cppfront -h`, there's
```output
Additional dynamic safety checks and contract information
  -a[dd-source-info]      Enable source_location information for contract checks
```

Though I've never seen it work.
I think it might only work with MSVC for now.

**Reply by gregmarr** [reply]
> I think it might only work with MSVC for now.

Yes, sadly.

    //  If this implementation doesn't support source_location yet, disable it
    #include <version>
    #if !defined(_MSC_VER) && !defined(__cpp_lib_source_location)
        #undef CPP2_USE_SOURCE_LOCATION
    #endif

**Reply by gregmarr** [reply]
<html>
<body>


std::source_location | P1208R6 | 11 | 15*(partial)16 | 19.29 (16.10)*
-- | -- | -- | -- | --
-- | -- | GCC libstdc++ | Clang libc++ | MSVC STL



</body>
</html>

[std::source_location](https://en.cppreference.com/w/cpp/utility/source_location)

**Reply by JohelEGP** [reply]
That needs updating.
Clang and GCC work: <https://compiler-explorer.com/z/bEfn94nvs>.
From <https://en.cppreference.com/w/cpp/compiler_support>:

C++20 feature | Paper(s) | GCC libstdc++ | Clang libc++ | MSVC STL | Apple Clang* |
-- | -- | -- | -- | -- | -- |
[std::source_location](https://en.cppreference.com/w/cpp/utility/source_location) |	[P1208R6](https://wg21.link/P1208R6) | 11 | 15*(partial)16 | 19.29 (16.10)* |   |

**Reply by gregmarr** [reply]
> From https://en.cppreference.com/w/cpp/compiler_support:

I think we both posted at about the same time.  :)

**Comment by gregmarr** [comment]
```Bounds safety violation in file.cpp:123: out-of-bounds access attempt detected```

For MSVC, with `-a`, you should get this:

```
filename.cpp2(123): Bounds safety violation: out-of-bounds access attempt detected
```
